src/backends/bitbucket/API.js

```diff
@@ -0,0 +1,191 @@
+      return baseHeader;
+    }
+
+    return baseHeader;
+  }
+
+  parseJsonResponse(response) {
+    return response.json().then((json) => {
+      if (!response.ok) {
+        return Promise.reject(json);
+      }
+
+      return json;
+    });
+  }
+
+  urlFor(path, options) {
+    const cacheBuster = new Date().getTime();
+    const params = [`ts=${cacheBuster}`];
+    if (options.params) {
+      for (const key in options.params) {
+        params.push(`${ key }=${ encodeURIComponent(options.params[key]) }`);
+      }
+    }
+    if (params.length) {
+      path += `?${ params.join("&") }`;
+    }
+    return this.api_root + path;
+  }
+
+  request(path, options = {}) {
+    const headers = this.requestHeaders(options.headers || {});
+    const url = this.urlFor(path, options);
+    let responseStatus;
+    return fetch(url, { ...options, headers }).then((response) => {
+      responseStatus = response.status;
+      const contentType = response.headers.get("Content-Type");
+      if (contentType && contentType.match(/json/)) {
+        return this.parseJsonResponse(response);
+      }
+      return response.text();
+    })
+    .catch((error) => {
+      throw new APIError(error.message, responseStatus, 'Bitbucket');
+    });
+  }
+
+  //formrequest uses different headers for a form request vs normal json request
+  formRequest(path, options = {}) {
+    const headers = this.formRequestHeaders(options.headers || {});
+    const url = this.urlFor(path, options);
+    let responseStatus;
+    return fetch(url, { ...options, headers }).then((response) => {
+      responseStatus = response.status;
+      const contentType = response.headers.get("Content-Type");
+      if (contentType && contentType.match(/json/)) {
+        return this.parseJsonResponse(response);
+      }
+      return response.text();
+    })
+    .catch((error) => {
+      throw new APIError(error.message, responseStatus, 'Bitbucket');
+    });
+  }
+
+  readFile(path, sha, branch = this.branch) {
+    const cache = sha ? LocalForage.getItem(`gh.${ sha }`) : Promise.resolve(null);
+    return cache.then((cached) => {
+      if (cached) { return cached; }
+
+      return this.request(`${ this.repoURL }/src/${ branch }/${ path }`, {
+        cache: "no-store",
+      }).then((result) => {
+        if (sha) {
+          LocalForage.setItem(`gh.${ sha }`, result);
+        }
+        return result;
+      });
+    });
+  }
+
+  listFiles(path) {
+    return this.request(`${ this.repoURL }/src/${ this.branch }/${ path }`, {})
+    .then(files => {
+      if (!Array.isArray(files.values)) {
+        throw new Error(`Cannot list files, path ${path} is not a directory`);
+      }
+      return files.values;
+    })
+    .then(files => {
+      return files.filter(file => file.type === "commit_file")
+    });
+  }
+
+  persistFiles(entry, mediaFiles, options) {
+    const uploadPromises = [];
+    const files = mediaFiles.concat(entry);
+
+    files.forEach((file) => {
```

Comment:
@zanedev This has been discussed before, and there are obviously pros and cons to both sides, but I wonder if it would be best to upload all the media files first, then upload the entry. That way, if there is a problem with one of the media files, the entry would not be broken. But that is an opinion, some people may prefer to have the content saved first.